src/modules/replay/replay_main.cpp

```diff
@@ -752,11 +731,273 @@ void Replay::task_main()
 
 		//TODO: should we close the log file & exit (optionally, by adding a parameter -q) ?
 	}
+
+	onExitMainLoop();
+}
+
+void Replay::readTopicDataToBuffer(const Subscription &sub, std::ifstream &replay_file)
+{
+	const size_t msg_read_size = sub.orb_meta->o_size_no_padding;
+	const size_t msg_write_size = sub.orb_meta->o_size;
+	_read_buffer.reserve(msg_write_size);
+	replay_file.seekg(sub.next_read_pos + (streamoff)(ULOG_MSG_HEADER_LEN + 2)); //skip header & msg id
+	replay_file.read((char *)_read_buffer.data(), msg_read_size);
+}
+
+bool Replay::handleTopicUpdate(Subscription &sub, void *data, std::ifstream &replay_file)
+{
+	return publishTopic(sub, data);
+}
+
+uint64_t Replay::handleTopicDelay(uint64_t next_file_time, uint64_t timestamp_offset)
+{
+
+	const uint64_t publish_timestamp = next_file_time + timestamp_offset;
+
+	//wait if necessary
+	uint64_t cur_time = hrt_absolute_time();
+
+	// if some topics have a timestamp smaller than the log file start, publish them immediately
+	if (cur_time < publish_timestamp && next_file_time > _file_start_time) {
+		usleep(publish_timestamp - cur_time);
+	}
+
+	return publish_timestamp;
+}
+
+bool Replay::publishTopic(Subscription &sub, void *data)
+{
+	bool published = false;
+
+	if (sub.orb_advert) {
+		orb_publish(sub.orb_meta, sub.orb_advert, data);
+		published = true;
+
+	} else {
+		if (sub.multi_id == 0) {
+			sub.orb_advert = orb_advertise(sub.orb_meta, data);
+			published = true;
+
+		} else {
+			// make sure the other instances are advertised already so that we get the correct instance
+			bool advertised = false;
+
+			for (const auto &subscription : _subscriptions) {
+				if (subscription.orb_meta) {
+					if (strcmp(sub.orb_meta->o_name, subscription.orb_meta->o_name) == 0 &&
+					    subscription.orb_advert && subscription.multi_id == sub.multi_id - 1) {
+						advertised = true;
+					}
+				}
+			}
+
+			if (advertised) {
+				int instance;
+				sub.orb_advert = orb_advertise_multi(sub.orb_meta, data, &instance, ORB_PRIO_DEFAULT);
+				published = true;
+			}
+		}
+	}
+
+	if (published) {
+		++sub.publication_counter;
+	}
+
+	return published;
+}
+
+bool ReplayEkf2::handleTopicUpdate(Subscription &sub, void *data, std::ifstream &replay_file)
+{
+	if (sub.orb_meta == ORB_ID(ekf2_timestamps)) {
+		ekf2_timestamps_s ekf2_timestamps;
+		memcpy(&ekf2_timestamps, data, sub.orb_meta->o_size);
+
+		if (!publishEkf2Topics(ekf2_timestamps, replay_file)) {
+			return false;
+		}
+
+		px4_pollfd_struct_t fds[1];
+		fds[0].fd = _vehicle_attitude_sub;
+		fds[0].events = POLLIN;
+		// wait for a response from the estimator
+		int pret = px4_poll(fds, 1, 1000);
+
+		// introduce some breaks to make sure the logger can keep up
+		if (++_topic_counter == 50) {
+			usleep(1000);
```

Comment:
Yes, the replay does not explicitly synchronize with the logger, so potentially it could miss samples during replay. Let's see how well this works.